File: packages/sdk-py/dna/adapters/filesystem/cache.py
"""FilesystemCache — CachePort backed by .dna-cache/ directories."""
from __future__ import annotations

import logging
import shutil
from pathlib import Path
from typing import Any

import aiofiles
import yaml

from dna._yaml import safe_load
from dna.kernel.protocols import CacheItem
from dna.kernel.bundle.handle import FilesystemBundleHandle

logger = logging.getLogger(__name__)


class FilesystemCache:
    """Stores and loads cached dependencies from .dna-cache/<scope>/."""

    def __init__(self, base_dir: str | Path) -> None:
        self._base = Path(base_dir).resolve().parent / ".dna-cache"

    def _contained(self, path: Path) -> Path:
        """Resolve ``path`` and refuse it unless it stays under ``.dna-cache``.

        The same second layer ``FilesystemSource._contained`` is, applied here
        because the segments this adapter joins reach it from FURTHER AWAY than
        anything the source adapter sees. ``scope`` and ``key`` are tame —
        every ``ResolverPort.cache_key`` sanitises its uri through
        ``re.sub(r"[^a-zA-Z0-9_-]", "-", …)``, so a key cannot carry a ``/`` or
        a dot at all. ``item.kind`` and ``item.name`` are not: ``HttpResolver``
        reads both straight out of REMOTE JSON (``raw["kind"]``,
        ``raw["metadata"]["name"]``), and they land in a ``shutil.copytree``
        destination — and in an ``rmtree`` of that destination first, which is
        the worse half.

        Nothing is known to abuse it today; this is the layer that means a
        hostile or merely careless registry cannot, rather than an incident
        being cleaned up. Refuses with ``PathEscapesStoreRoot``, the same
        refusal the source adapter raises, so a face relays it identically.
        """
        resolved = Path(path).resolve()
        try:
            resolved.relative_to(self._base.resolve())
        except ValueError as exc:
            from dna.kernel.errors import PathEscapesStoreRoot

            raise PathEscapesStoreRoot(
                f"FilesystemCache refused a path that leaves its cache root: "
                f"{str(resolved)!r} is not under {str(self._base)!r}. The "
                f"scope, the resolver cache key and the resolved item's kind + "
                f"name are all PATH COMPONENTS here, and the last two come "
                f"from the remote payload. Fix the resolver's output; do not "
                f"widen this check."
            ) from exc
        return resolved

    async def has(self, scope: str, key: str) -> bool:
        return self._contained(self._base / scope / key).exists()

    async def store(self, scope: str, key: str, items: list[CacheItem]) -> None:
        dest_base = self._contained(self._base / scope / key)
        dest_base.mkdir(parents=True, exist_ok=True)
        for item in items:
            # When kind is known, organize by kind (skills/, souls/)
            # When kind is "" (scanner determines later), store flat by name
            if item.kind:
                sub_dir = self._contained(dest_base / (item.kind.lower() + "s"))
            else:
                sub_dir = dest_base
            # ``item.name`` is remote-derived on the HTTP resolver — contained
            # BEFORE the rmtree, not only before the copy.
            dest = self._contained(sub_dir / item.name)
            if dest.exists():
                shutil.rmtree(dest)
            dest.parent.mkdir(parents=True, exist_ok=True)
            shutil.copytree(item.content_path, dest)

    async def load_key(
        self, scope: str, key: str, readers: list | None = None,
    ) -> list[dict[str, Any]]:
        key_dir = self._contained(self._base / scope / key)
        if not key_dir.exists():
            return []
        documents: list[dict[str, Any]] = []
        await self._read_tree(key_dir, readers or [], documents)
        return documents

    async def load_all(
        self, scope: str, readers: list | None = None,
    ) -> list[dict[str, Any]]:
        scope_dir = self._contained(self._base / scope)
        if not scope_dir.exists():
            return []

        documents: list[dict[str, Any]] = []
        readers = readers or []
        await self._read_tree(scope_dir, readers, documents)
        return documents

    async def _read_tree(
        self, directory: Path, readers: list,
        documents: list[dict[str, Any]],
    ) -> None:
        """Recursively read directories for bundles and YAMLs."""
        for subdir in sorted(directory.iterdir()):
            if not subdir.is_dir():
                continue
            # Try readers first
            matched = False
            bundle = FilesystemBundleHandle(subdir)
            for reader in readers:
                try:
                    if reader.detect(bundle):
                        doc = reader.read(bundle)
                        if isinstance(doc, dict) and "kind" in doc:
                            documents.append(doc)
                        matched = True
                        break
                except Exception as e:
                    logger.warning("Reader error on %s: %s", subdir, e)
            if matched:
                continue
            # Check for YAML files with kind:
            has_yaml = False
            for yf in sorted(subdir.glob("*.yaml")):
                try:
                    async with aiofiles.open(yf, "r") as f:
                        raw = await f.read()
                    content = safe_load(raw)
                    if isinstance(content, dict) and "kind" in content:
                        documents.append(content)
                        has_yaml = True
                except yaml.YAMLError:
                    pass
            # Recurse deeper if nothing found at this level
            if not has_yaml:
                await self._read_tree(subdir, readers, documents)

File: packages/sdk-ts/src/adapters/filesystem/cache.ts
/**
 * FilesystemCache — CachePort backed by .dna-cache/ directories.
 *
 * 1:1 parity with Python dna.v3.adapters.filesystem.cache.
 */

import { access, cp, mkdir, readFile, readdir, rm, stat } from "node:fs/promises";
import { resolve, join, dirname } from "node:path";

async function pathExists(p: string): Promise<boolean> {
  try { await access(p); return true; } catch { return false; }
}
import yaml from "js-yaml";
import type { CacheItem, CachePort, ReaderPort } from "../../kernel/protocols.js";
import { FilesystemBundleHandle } from "../../kernel/bundle-handle.js";

export class FilesystemCache implements CachePort {
  readonly baseDir: string;
  private readonly _cacheDir: string;

  constructor(baseDir: string) {
    this.baseDir = resolve(baseDir);
    this._cacheDir = join(dirname(this.baseDir), ".dna-cache");
  }

  async has(scope: string, key: string): Promise<boolean> {
    return pathExists(join(this._cacheDir, scope, key));
  }

  async store(scope: string, key: string, items: CacheItem[]): Promise<void> {
    const destBase = join(this._cacheDir, scope, key);
    await mkdir(destBase, { recursive: true });

    for (const item of items) {
      const subDir = item.kind
        ? join(destBase, item.kind.toLowerCase() + "s")
        : destBase;
      const dest = join(subDir, item.name);

      if (await pathExists(dest)) {
        await rm(dest, { recursive: true, force: true });
      }
      await mkdir(dirname(dest), { recursive: true });
      await cp(item.contentPath, dest, { recursive: true });
    }
  }

  async loadKey(
    scope: string,
    key: string,
    readers?: ReaderPort[],
  ): Promise<Record<string, unknown>[]> {
    const keyDir = join(this._cacheDir, scope, key);
    if (!(await pathExists(keyDir))) return [];

    const documents: Record<string, unknown>[] = [];
    await this._readTree(keyDir, readers ?? [], documents);
    return documents;
  }

  async loadAll(
    scope: string,
    readers?: ReaderPort[],
  ): Promise<Record<string, unknown>[]> {
    const scopeDir = join(this._cacheDir, scope);
    if (!(await pathExists(scopeDir))) return [];

    const documents: Record<string, unknown>[] = [];
    await this._readTree(scopeDir, readers ?? [], documents);
    return documents;
  }

  private async _readTree(
    directory: string,
    readers: ReaderPort[],
    documents: Record<string, unknown>[],
  ): Promise<void> {
    const entries = (await readdir(directory)).sort();
    for (const entry of entries) {
      const full = join(directory, entry);
      if (!(await stat(full)).isDirectory()) continue;

      let matched = false;
      const handle = new FilesystemBundleHandle(full);
      for (const reader of readers) {
        try {
          if (await reader.detect(handle)) {
            const doc = await reader.read(handle);
            if (
              doc != null &&
              typeof doc === "object" &&
              "kind" in doc
            ) {
              documents.push(doc);
            }
            matched = true;
            break;
          }
        } catch {
          // Skip reader errors
        }
      }
      if (matched) continue;

      let hasYaml = false;
      const subEntries = (await readdir(full)).sort();
      for (const sub of subEntries) {
        if (!sub.endsWith(".yaml") && !sub.endsWith(".yml")) continue;
        const yf = join(full, sub);
        if (!(await stat(yf)).isFile()) continue;
        try {
          const content = yaml.load(await readFile(yf, "utf-8"));
          if (
            content != null &&
            typeof content === "object" &&
            "kind" in (content as Record<string, unknown>)
          ) {
            documents.push(content as Record<string, unknown>);
            hasYaml = true;
          }
        } catch {
          // Skip unparseable YAML
        }
      }

      if (!hasYaml) {
        await this._readTree(full, readers, documents);
      }
    }
  }
}
